src/ImageSharp/Formats/ImageDecoder{T}.cs

```diff
@@ -0,0 +1,119 @@
+    /// The base class for all image decoders.
+    /// </summary>
+    /// <typeparam name="T">The type of specialized decoder options.</typeparam>
+    public abstract class ImageDecoder<T> : IImageInfoDetector, IImageDecoder
+        where T : class, ISpecializedDecoderOptions, new()
+    {
+        /// <inheritdoc/>
+        public IImageInfo Identify(DecoderOptions options, Stream stream, CancellationToken cancellationToken)
+        {
+            T specializedOptions = new() { GeneralOptions = options };
+            return this.IdentifySpecialized(specializedOptions, stream, cancellationToken);
+        }
+
+        /// <inheritdoc/>
+        public Image<TPixel> Decode<TPixel>(DecoderOptions options, Stream stream, CancellationToken cancellationToken)
+            where TPixel : unmanaged, IPixel<TPixel>
+        {
+            T specializedOptions = new() { GeneralOptions = options };
+            Image<TPixel> image = this.DecodeSpecialized<TPixel>(specializedOptions, stream, cancellationToken);
+
+            Resize(options, image);
+
+            return image;
+        }
+
+        /// <inheritdoc/>
+        public Image Decode(DecoderOptions options, Stream stream, CancellationToken cancellationToken)
+        {
+            T specializedOptions = new() { GeneralOptions = options };
+            Image image = this.DecodeSpecialized(specializedOptions, stream, cancellationToken);
+
+            Resize(options, image);
+
+            return image;
+        }
+
+        /// <summary>
+        /// Reads the raw image information from the specified stream.
+        /// </summary>
+        /// <param name="options">The specialized decoder options.</param>
+        /// <param name="stream">The <see cref="Stream"/> containing image data.</param>
+        /// <param name="cancellationToken">The token to monitor for cancellation requests.</param>
+        /// <returns>The <see cref="IImageInfo"/> object.</returns>
+        /// <exception cref="ImageFormatException">Thrown if the encoded image contains errors.</exception>
+        public abstract IImageInfo IdentifySpecialized(T options, Stream stream, CancellationToken cancellationToken);
+
+        /// <summary>
+        /// Decodes the image from the specified stream to an <see cref="Image{TPixel}"/> of a specific pixel type.
+        /// </summary>
+        /// <typeparam name="TPixel">The pixel format.</typeparam>
+        /// <param name="options">The specialized decoder options.</param>
+        /// <param name="stream">The <see cref="Stream"/> containing image data.</param>
+        /// <param name="cancellationToken">The token to monitor for cancellation requests.</param>
+        /// <returns>The <see cref="Image{TPixel}"/>.</returns>
+        /// <exception cref="ImageFormatException">Thrown if the encoded image contains errors.</exception>
+        public abstract Image<TPixel> DecodeSpecialized<TPixel>(T options, Stream stream, CancellationToken cancellationToken)
```

Comment:
`DecodeSpecialized<T>` doesn't look like a real abstraction, it's rather a pattern. We never call it though the abstract base method. We could just define the methods in decoder subclasses without overriding anything.

I think that using inheritance and generics to enforce a pattern is a wrong design decision. Can't refer to any design guideline paragraphs on this, but I'm not aware of any precedents in the BCL aspnet, or other libraries.

Reply:
I'm not following... The abstract base class calls it from `Decode`.

https://github.com/SixLabors/ImageSharp/pull/2180/files/085ff1ae465e67e27a80dae0601a56ae3149c66d#diff-7618afaee324d1c3d35802c1d55049a6226f471b8f73295c454e80d8f01e5290R29-R30

Reply:
I missed this. Still not big fan of the design, since fact the specialized method (and as a consequence the abstraction) is unnecessary for many formats, see next comment.

Reply:
I'll see what I can do but it's tricky. There's no concrete implementation for the core decoders so I can't implement a single abstract way to ensure the decoder honors the resize operation without reimplementing each time.